openlibrary/plugins/openlibrary/js/SearchBar.js

```diff
@@ -0,0 +1,298 @@
+        }
+        url += `&mode=${SearchUtils.mode.read()}`;
+        return url;
+    }
+
+    /**
+     * Prepare an unprocessed query for book searching
+     * @param {String} q
+     * @return {String}
+     */
+    static marshalBookSearchQuery(q) {
+        if (q && q.indexOf(':') == -1 && q.indexOf('"') == -1) {
+            q = `title: "${q}"`;
+        }
+        return q;
+    }
+
+    /** Setup event listeners for autocompletion */
+    initAutocompletionLogic() {
+        // searches should be cancelled if you click anywhere in the page
+        $(document.body).on('click', this.clearAutocompletionResults.bind(this));
+        // but clicking search input should not empty search results.
+        this.$input.on('click', false);
+
+        this.$input.on('keyup', debounce(event => {
+            // ignore directional keys and enter for callback
+            if (![13,37,38,39,40].includes(event.keyCode)) {
+                this.renderAutocompletionResults();
+            }
+        }, 500, false));
+
+        this.$input.on('focus', debounce(event => {
+            event.stopPropagation();
+            this.renderAutocompletionResults();
+        }, 300, false));
+    }
+
+    /** Cleans up and performs the query, then update the autocomplete results */
+    renderAutocompletionResults() {
+        let q = this.$input.val();
+        if (q === '' || !(this.facetEndpoint in RENDER_AUTOCOMPLETE_RESULT)) {
+            return;
+        }
+        if (this.facet.read() === 'title') {
+            q = SearchBar.marshalBookSearchQuery(q);
+        }
+
+        this.$results.css('opacity', 0.5);
+        $.getJSON(this.composeSearchUrl(q, true, 10), data => {
+            const renderer = RENDER_AUTOCOMPLETE_RESULT[this.facetEndpoint];
+            this.$results.css('opacity', 1);
+            this.clearAutocompletionResults();
+            for (let d in data.docs) {
+                this.$results.append(renderer(data.docs[d]));
+            }
+        });
+    }
+
+    clearAutocompletionResults() {
+        this.$results.empty();
+    }
+
+    /**
+     * Updates the UI to match after the facet is changed
+     * @param {String} newFacet
+     */
+    handleFacetValueChange(newFacet) {
+        // update the UI
+        this.$facetSelect.val(newFacet);
+        const text = this.$facetSelect.find('option:selected').text();
+        $('header#header-bar .search-facet-value').html(text);
+
+        // Get new results
+        if (this.$input.is(':focus')) {
+            this.renderAutocompletionResults();
+        }
+    }
+
+    /**
+     * Handles changes to the facet from the UI
+     * @param {JQuery.Event} event
+     */
+    handleFacetSelectChange(event) {
+        const newFacet = this.$facetSelect.val();
+        // We don't want to persist advanced becaues it behaves like a button
+        if (newFacet == 'advanced') {
+            event.preventDefault();
+            window.location.assign('/advancedsearch');
```

Comment:
What if this just ran a function passed in to the constructor?
e.g.

```
new SearchBar( {
  onFacetAdvanceSelect: function ( newFace ) { window.location.assign('/advancedsearch'); }
 } )
```

Reply:
I agree it would be more testable, and reusable, but not readable. It makes the constructor look unnecessarily complex, since this is something that is unlikely to change. As per reusability, I'd argue this is still reusable, just not super customizeable. And since this is only used in one place, adding that extra config is a violation of "You aren't gonna need it". So I don't think the testability benefits are worth the loss in readability or the violation of YAGNI. So closing as WONT FIX